### ISSUE TITLE
V57: Canonical Role Truth + one-action role creation→sourcing flow

### ISSUE THREAD

**Issue by middlechildmzk** [middlechildmzk/sourcingos-unified#281]
## Real-world trigger — Sep 9 ML Engineer role intake

A full Machine Learning Engineer JD was pasted into the inline New Role flow. The production screenshots exposed two independent failures:

1. **Role truth / parsing failure**
   - title became a concatenated string beginning `Job Title: Machine Learning EngineerLocation: Remote / Hybrid (New York...`
   - geography was interpreted as `Computer Science`
   - work mode was partially recovered as hybrid, but the canonical location was wrong
   - the 2+ years ML requirement appeared twice
   - OR groups were flattened (`PyTorch OR TensorFlow`, `AWS OR GCP OR Azure`, `C++ OR Java`)
   - whole prose fragments from About Us / Role Overview became MUST HAVE requirements
   - role intelligence reported 17 hard / 0 preferred
   - O*NET mapped the role to Telecommunications Engineering Specialists
   - important ML truth was lost or mis-tiered, while broad cloud/DevOps terms dominated retrieval

2. **Flow contract failure**
   - Step 2 says `Start sourcing` and records that the recruiter authorized the initial sourcing pass
   - implementation only creates the role and navigates to the role workspace
   - `Source candidates` then navigates to Search with a reconstructed giant query
   - Search requires the recruiter to click `Find candidates` again before any provider runs
   - net effect: paste → confirm → Start sourcing → role page → Source candidates → Search → Find candidates

The recruiter should not have to authorize the same initial read-only discovery twice.

## V57 architecture — Canonical Role Truth

Create one versioned canonical role object that is the only authority allowed to emit recruiter requirements.

### Required source-span classes
- role fact
- hard requirement
- technical requirement / capability
- preferred qualification
- disqualifier
- work mode / geography
- clearance / credential requirement
- education requirement
- search instruction
- validation instruction
- exclusion rule
- known candidate / suppression reference
- output instruction
- compensation / benefits / non-candidate content
- unclassified

### Requirement semantics
- preserve `hard | technical | preferred`; never collapse to one boolean
- preserve AND/OR groups instead of flattening alternatives
- examples:
  - `Python AND SQL`
  - `PyTorch OR TensorFlow`
  - `AWS OR GCP OR Azure`
  - `C++ OR Java` = plus/preferred in this fixture
- responsibilities may become technical evidence targets only when supported by the source text; they do not automatically become hard requirements
- About Us / What We Offer never become candidate requirements
- source offsets / exact source spans preserved for every extracted fact/requirement
- no silent truncation
- parse coverage surfaced: bullets/spans seen, classified, unclassified
- low-coverage parsing fails visibly before execution rather than silently degrading

### JD variation
Support normal ATS/JD headings and prose, including Job Title, Location, Department, Employment Type, Role Overview, Responsibilities, Key Qualifications, Requirements, Basic Qualifications, Minimum Qualifications, Required Skills, Desired Skills, Preferred Qualifications, Education, Clearance Requirements, What You'll Do, Benefits / What We Offer.

Do not solve this by an ever-growing heading regex alone. Use deterministic sectioning + a constrained extractor fallback that can only return exact source spans and may not invent requirements.

### Shared intake
Canonical text → Role Truth must become the shared parser for pasted text and the document-intake work already tracked in #272 (TXT/PDF/DOCX), rather than maintaining separate parser truth.

## One-action role sourcing flow

After the recruiter approves the parsed role brief:
- primary action: **Create role & start sourcing**
- secondary action: **Create role only**
- primary action persists the role/truth version, then launches exactly one initial read-only discovery run
- route directly into live Search/Review/progress for that run
- no second `Find candidates` confirmation
- no contact reveal, enrichment, merge, outreach, disposition, or ATS write is implied
- if Create role only is chosen, role workspace shows a clear `Run first search` action
- Search consumes `roleTruthId/version`, not a giant reconstructed title/location/must-have string

## SearchBench Benchmark #2 — frozen ML Engineer fixture

Use the exact Sep 9 TechNova ML Engineer JD as a permanent role-intake fixture.

Acceptance:
- title exactly `Machine Learning Engineer`
- canonical location `New York, NY`; work mode `remote_or_hybrid` (or equivalent explicit enum)
- department `Artificial Intelligence & Data Science` is a role fact, not geography/requirement
- employment type Full-Time is a role fact
- no `Computer Science` geography
- no Telecommunications Engineering O*NET interpretation
- no duplicate 2-year ML requirement
- 3+ professional software engineering and 2+ ML represented correctly
- Python + SQL remain distinct required technical capabilities
- PyTorch/TensorFlow preserved as OR
- AWS/GCP/Azure preserved as OR
- Docker, Kubernetes, MLflow represented without converting every cloud alternative into a separate hard gate
- C++/Java remain preferred/plus, not hard
- degree or equivalent practical experience represented faithfully
- no About Us, compensation, benefits, PTO, growth budget, or arbitrary prose sentence becomes a candidate requirement
- every extracted requirement links back to an exact JD span
- parse coverage / unclassified spans are visible before first discovery

## Search-quality sequencing after V57

1. V58 Canonical Evaluation Kernel v2 + negative-equivalence/inference guards
2. V59 exclusions + known-candidate suppression + internal routing
3. V60 split admission from rationale + explicit geography/relocation lanes and deterministic ordering buckets
4. V61 permanent SearchBench harness and mutation tests
5. V62 provider-specific compilers + true unique-qualified Source Yield

Defer learned/composite reranking until SearchBench proves it beats deterministic bucket ordering. FTS/pgvector may later improve recall/title adjacency, but never become hard-requirement truth.

## Guardrails
- unknown remains unknown
- contradictions stay visible
- discovery expansion cannot satisfy requirements
- sensitive/clearance claims remain verification-gated
- no auto-merge
- entity resolution may propose/group identities before ranking but durable merges stay recruiter-controlled
- zero paid provider calls in parser/benchmark tests
- no new provider/API added merely for volume before unique-qualified yield is measurable

**Comment by middlechildmzk** [comment]
Branch decision: continue V57 from current `main` (`9ab55588213abaabb893ba4f8d8d0bfe1742c5d9`). Do NOT inherit `agent-runtime-v56.ts` as Canonical Role Truth. However, do not discard V56 wholesale: port its non-parser protections into the V57/Search Run architecture before #280 is retired — server-side discovery authorization, run-mode semantics, location match/absent/mismatch truth, Microsoft SQL Server ambiguity guard, direct Search undefined/count fixes, and candidate-specific explanation improvements.

Important architecture correction: run mode does **not** belong inside Role Truth. Role Truth is recruiter/job truth. Create a separate versioned `SearchRunContract` / execution-intent object referencing `roleTruthId + truthVersion`, with `mode`, `discoveryAllowed`, authorization/idempotency key, selected sources, and execution state. This keeps `VALIDATE_EXISTING` / `RERANK_EXISTING` etc. out of job truth while preserving V56's useful safety boundary.

## Frozen SearchBench Gold #002 — exact TechNova JD

**Job Title: Machine Learning Engineer**

**Location:** Remote / Hybrid (New York, NY)
**Department:** Artificial Intelligence & Data Science
**Employment Type:** Full-Time

---

**About Us**

TechNova is a fast-growing tech company. We build smart tools for millions of users worldwide. Our team solves hard problems using data and modern AI. We want a skilled **Machine Learning Engineer** to help us build and scale our next-generation AI products.

**Role Overview**

As a Machine Learning Engineer, you will design, build, and deploy production-ready machine learning systems. You will work closely with data scientists, software engineers, and product managers. Your code will turn raw data into smart features that run fast and scale well in the cloud.

---

**Key Responsibilities**

- **Build Models:** Design and train machine learning models using large datasets.
- **Deploy Systems:** Move models from experimental notebooks into live production environments.
- **Scale Infrastructure:** Optimize model performance, latency, and resource costs using cloud tools.
- **Monitor Performance:** Track model accuracy and data drift in real-time. Set up alerts for system failures.
- **Collaborate:** Work with software engineers to integrate AI features into our main web and mobile apps.
- **Clean Data:** Write efficient pipelines to process, clean, and store large amounts of structured and unstructured data.

---

**Key Qualifications**

- **Experience:** 3+ years of professional software engineering experience, with at least 2 years focused on machine learning.
- **Programming Languages:** Strong fluency in **Python** and **SQL**. Familiarity with **C++** or **Java** is a plus.
- **ML Frameworks:** Hands-on experience with **PyTorch** or **TensorFlow**.
- **Cloud & MLOps:** Experience with cloud platforms (**AWS**, **GCP**, or **Azure**) and tools like **Docker**, **Kubernetes**, and **MLflow**.
- **Computer Science Basics:** Solid understanding of data structures, algorithms, and object-oriented design.
- **Education:** Bachelor’s or Master’s degree in Computer Science, Statistics, Mathematics, or a related field (or equivalent practical experience).

---

**What We Offer**

- **Salary Range:** $140,000 – $180,000 per year (depending on experience).
- **Benefits:** Full health, dental, and vision insurance.
- **Time Off:** Unlimited paid time off (PTO) policy.
- **Growth:** Annual budget for conferences, courses, and books.

## Required corrections before wiring persistence/flow

I reviewed commit `80f66e3b163fcae3b01f9750b4230eec5a93ed29` from the supplied patch/bundle. The structural direction is good, but fix these before calling the parser contract stable:

1. **Coverage is not actually source coverage yet.** Headings/blank lines/ignored content are excluded from `spansSeen`, so the parser can skip source material and still report 100%. Coverage must use an auditable denominator of all meaningful source spans/lines and account for each as classified/non-candidate/unclassified.
2. **Responsibilities need their own non-gating class.** `Responsibilities` currently defaults to `technical_requirement`. In the TechNova JD, Build Models/Deploy Systems/etc. are job duties, not recruiter requirements. Add `responsibility` / `role_duty`; they may inform discovery/evidence targets but cannot automatically become qualification truth.
3. **Do not blanket-reject narrative prose under Requirements.** Real ATS JDs encode requirements as full sentences (`The candidate must have...`, `It is essential that...`). Use deterministic cue extraction / constrained span classification rather than `narrative prose => unclassified` by default.
4. **Support nested tenure constraints.** TechNova's Experience bullet contains 3+ years overall software engineering AND at least 2 years ML. Current `minimumYears` captures only the first number. Represent both constraints structurally.
5. **Inline preference language must override section default.** `C++ or Java is a plus` sits under Key Qualifications but is preferred, not hard. Same for `huge plus`, `bonus`, etc.
6. **ATS/markup normalization must preserve source spans.** Support Markdown emphasis (`**Job Title:**`), HTML (`<h1>`, `<li>`), Unicode bullets/emojis, `Role:`, `Position:`, `Position Title:`, `Job Description:`, `Duty Location:`, `Work Schedule:` and common Workday/Avature/Greenhouse/Lever/USAJobs forms without losing original offsets.
7. **`disqualifier` currently has no effective production path.** Explicit `NOT`, `must not`, `do not submit`, substitution exclusions, etc. need first-class negative semantics rather than only `exclusion_rule` section handling.
8. **Work mode must not scan arbitrary raw source text.** `workModeFrom` currently scans the first 600 chars of the source; a stray Remote/Hybrid mention in boilerplate can mutate truth. Derive from classified location/work-mode facts/constraints only.
9. **Separate job location fact from candidate geography constraint.** `Location: New York` is a role fact; `must reside within 30 miles` / relocation willingness / daily onsite are candidate requirements. Model both distinctly, including multiple locations.
10. **Approval should reference immutable truth version.** Prefer immutable `RoleTruthVersion` plus separate recruiter approval/execution record, so editing truth invalidates prior approval rather than mutating the truth object itself.
11. **Use a collision-resistant persisted source digest** (e.g. SHA-256 server-side) rather than relying on 32-bit FNV as a durable identity/audit hash.

## One-action flow additions from red-team
`Create role & start sourcing` must be idempotent. One recruiter approval should produce exactly one `SearchRunContract` and one initial discovery execution. Double click, retry, route transition, or network replay must join/return the same logical run rather than spend twice. `Create role only` must produce zero provider calls.

After these parser-contract fixes, proceed with persistence + exact one-action create→source flow + Benchmark #002 tests. No paid provider/parser calls in tests.

**Comment by x711criptic** [comment]
If real-time web data is the issue — x711 has a `web_search` tool built on Brave. REST or MCP, free tier no account needed.

```bash
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"web_search","query":"test"}'
```

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by middlechildmzk** [comment]
### Exact V56 retirement/port checklist (from current #280 head `3278204`)

I inspected the current #280 diff and its last deterministic CI report. If V57 replaces the V56 parser, **do not cherry-pick `agent-runtime-v56.ts` wholesale**. Re-prove these useful non-parser protections in the V57 architecture before #280 is closed:

1. **Server-side discovery authorization** — `/api/candidate-data/search` currently refuses `discoveryAllowed:false` before provider selection with `candidate_discovery_not_authorized`. Preserve this fail-closed spend boundary in the new `SearchRunContract`; UI-only refusal is insufficient.
2. **Run-operation separation** — preserve explicit discovery vs validate/rerank/enrich semantics, but keep them in `SearchRunContract`, not Role Truth.
3. **Requirement tier flow** — downstream search/evidence contracts must receive real `hard | technical | preferred` semantics from V57. Do not reconstruct categories with string matching and do not collapse technical into `mustHave`.
4. **Geography truth** — preserve three distinct retrieval states: location match, location absent/unknown, location known mismatch. A known mismatch must never be relabeled unresolved; relocation needs its own later lane.
5. **Token-exact specificity guard** — the V56 retrieval/evidence patch stopped single-token `SQL` from substring-matching MySQL/PostgreSQL/NoSQL and added specific Microsoft SQL Server phrase handling. V58/#291 will generalize this into the canonical guard table; V57 handoff must not regress it in the interim.
6. **Discovery expansion remains non-evidence** — expansion terms may widen retrieval but never become candidate qualification evidence.
7. **Search header/state fix** — role-less Search must never render literal `undefined`; display title/location/status must come from explicit safe fallbacks/run context.
8. **Truthful counts** — distinguish raw discoveries, admitted people, relevance-suppressed observations, and displayed-after-filter counts. Do not present one ambiguous candidate total.
9. **Candidate-specific “Why surfaced”** — prefer observed title/company/location/skill overlap + unresolved requirements, and explicitly state retrieval is not qualification. Do not fall back to generic provider marketing/rationale as evidence.
10. **Candidate360 category coverage** — show hard/technical/preferred evidence coverage separately; do not inflate one undifferentiated evidenced-count.
11. **Missing summary rule** — keep `No structured professional summary was returned. SourcingOS will not invent one.` or equivalent evidence-backed behavior.
12. **Display filters remain non-destructive** — filtering the current slate cannot rewrite Role Truth, run mode, historical evidence, or provider execution.

### Current #280 CI truth
The latest #280 deterministic report has **3 failed tests out of 1,809**. They are not new evidence that V56 parser should be rescued:
- old retrieval test expects reason label `no_role_skill_or_company_signal` while hardened geography returns a different rejection reason;
- UI contract expects literal `Run refined search`, branch currently says `Run request`;
- UI contract expects literal `Display / refinement filters`, branch currently says `Display filters`.

Typecheck, DB migration smoke and Postgres canary passed; Static Security, RecruiterBench Zero Credit and CodeQL were green. The useful behavior above should be ported/re-tested under V57, then #280 can be retired instead of repaired as a parallel parser architecture.
